### PR TITLE
Refactor internal engine

### PR DESCRIPTION
**Pull request by jasontedor**
This commit is a minor refactoring of internal engine to move hooks for generating sequence numbers into the engine itself. As such, we refactor tests that relied on this hook to use the new hook, and remove the hook from the sequence number service itself.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -2008,8 +2031,7 @@ public MergeStats getMergeStats() {
         return mergeScheduler.stats();
     }
 
-    @Override
-    public SequenceNumbersService seqNoService() {
+    public final SequenceNumbersService seqNoService() {
```
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: Just wondering: Is it possible to move `seqNoService` field to `Engine` make the `seqNoService()` method there final? The `SeqNoServiceSupplier` (or BiFunction) should then also be moved as an constructor parameter to `Engine`.

File: test/framework/src/main/java/org/elasticsearch/index/engine/TranslogHandler.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptyMap;
+import static org.elasticsearch.index.mapper.SourceToParse.source;
+
+public class TranslogHandler implements EngineConfig.TranslogRecoveryRunner {
```
Comment (martijnvg) [inline test/framework/src/main/java/org/elasticsearch/index/engine/TranslogHandler.java]: Maybe add class jdoc here about what the purpose of this implementation is?

**Review by martijnvg — COMMENTED** [review]
I left some comments.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 3]
```diff
@@ -145,18 +145,27 @@
     private final String historyUUID;
 
     public InternalEngine(EngineConfig engineConfig) {
+        this(engineConfig, InternalEngine::sequenceNumberService);
+    }
+
+    @FunctionalInterface
+    protected interface SeqNoServiceSupplier {
```
[the remark this answers is not in the document]
Comment (jasontedor) [reply core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: This is a good catch. This is a product of the fact that I initially started with the supplier being a function of more parameters (the shard ID, the allocation ID, etc.) but collapsed these into the two that are left and did not see until you pointed it out that we can go all the way and remove the interface. Thank you.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 1 above)
Comment (jasontedor) [reply core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I don't think we can do this cleanly right now. It means that the sequence number service has to be available immediately (to pass to the super constructor) but we are also trying to initialize the index writer at the same time that we initialize the sequence number service.

Comment (martijnvg) [reply core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I see, that makes it trickier. Thanks for explaining that.

**Review by martijnvg — APPROVED** [review]
From my side this LGTM.

**Review by s1monw — APPROVED** [review]
I am curious where this goes, the refactoring LGTM. I wonder if we should do this in master rather than in a branch. looks generic enough?

**Comment by jasontedor** [comment]
@s1monw I'm happy to pull this into master/6.x rather than a long-running feature branch, it will require  maintenance on the feature branch then. 😇

**Pull request merged**